src/vm/sampleprofiler.cpp

```diff
@@ -0,0 +1,145 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "common.h"
+#include "sampleprofiler.h"
+#include "hosting.h"
+#include "threadsuspend.h"
+
+bool SampleProfiler::s_profilingEnabled = false;
+Thread* SampleProfiler::s_pSamplingThread = NULL;
+#ifdef FEATURE_PAL
+TIMESPEC SampleProfiler::s_samplingRate = {0, 1000000}; // 1ms
+#endif
+
+// Synchronization of multiple callers occurs in EventPipe::Enable.
+void SampleProfiler::Enable()
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_ANY;
+        PRECONDITION(s_pSamplingThread == NULL);
+    }
+    CONTRACTL_END;
+
+    s_profilingEnabled = true;
+    s_pSamplingThread = SetupUnstartedThread();
+    if(s_pSamplingThread->CreateNewThread(0, ThreadProc, NULL))
+    {
+        // Start the sampling thread.
+        s_pSamplingThread->SetBackground(TRUE);
+        s_pSamplingThread->StartThread();
+    }
+    else
+    {
+        _ASSERT(!"Unable to create sample profiler thread.");
+    }
+}
+
+// Synchronization of multiple callers occurs in EventPipe::Disable.
+void SampleProfiler::Disable()
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    // The sampling thread will watch this value and exit
+    // when profiling is disabled.
+    s_profilingEnabled = false;
+
+    // Wait for the sampling thread to clean itself up.
+    while(s_pSamplingThread != NULL)
+    {
+        ClrSleepEx(0, FALSE);
+    }
+}
+
+DWORD WINAPI SampleProfiler::ThreadProc(void *args)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_TRIGGERS;
+        MODE_PREEMPTIVE;
+        PRECONDITION(s_pSamplingThread != NULL);
+    }
+    CONTRACTL_END;
+
+    // Complete thread initialization and start the profiling loop.
+    if(s_pSamplingThread->HasStarted())
+    {
+        // Switch to pre-emptive mode so that this thread doesn't starve the GC.
+        GCX_PREEMP();
+
+        while(s_profilingEnabled && !g_fEEShutDown)
```

Comment:
Fixed.  Yes, you're right.  The g_fEEShutDown was in there from some testing I was doing, and I don't believe it's needed.